File: web-demo/public/js/gaze.js
// Head-pose-aware webcam gaze, ported from the readsync Python pipeline
// (webcam/tracker.py and webcam/gaze.py). Pure functions plus a small model
// class, with no browser or MediaPipe dependency, so they are unit-tested in
// Node and reused unchanged in the browser. The MediaPipe FaceLandmarker is
// driven in app.js, which passes its result objects to signalsFromResult here.

import { ridgeFit } from "./linalg.js";

// Landmark indices in the 478-point model: (iris, corner, corner, upper lid,
// lower lid) for each eye.
export const RIGHT_EYE = [468, 33, 133, 159, 145];
export const LEFT_EYE = [473, 362, 263, 386, 374];

export function eyeFeature(landmarks, spec) {
  // Iris position in the eye's own frame, as fractions of width and height. Using
  // the eye's own axes makes the feature invariant to head roll.
  const [iris, a, b, up, lo] = spec;
  const ax = landmarks[a].x, ay = landmarks[a].y;
  const bx = landmarks[b].x, by = landmarks[b].y;
  const ix = landmarks[iris].x, iy = landmarks[iris].y;
  const axisX = bx - ax, axisY = by - ay;
  const width = Math.hypot(axisX, axisY) || 1e-6;
  const ux = axisX / width, uy = axisY / width;
  const fx = ((ix - ax) * ux + (iy - ay) * uy) / width;
  const cx = (ax + bx) / 2, cy = (ay + by) / 2;
  const height =
    Math.hypot(landmarks[lo].x - landmarks[up].x, landmarks[lo].y - landmarks[up].y) || 1e-6;
  const fy = ((ix - cx) * -uy + (iy - cy) * ux) / height;
  return [fx, fy];
}

export function eyeOpenness(landmarks, spec) {
  // Eye-aspect ratio (lid gap over eye width). Falls towards zero during a blink.
  const [, a, b, up, lo] = spec;
  const width =
    Math.hypot(landmarks[b].x - landmarks[a].x, landmarks[b].y - landmarks[a].y) || 1e-6;
  const height = Math.hypot(landmarks[lo].x - landmarks[up].x, landmarks[lo].y - landmarks[up].y);
  return height / width;
}

export function headPoseFromMatrix(data) {
  // Extract (yaw, pitch, roll) in radians from a 16-element 4x4 head-pose matrix
  // (FaceLandmarker facial transformation matrix, row-major). The exact convention
  // matters only for naming; the angles serve as head-orientation inputs and to
  // measure drift.
  const r = (i, j) => data[i * 4 + j];
  const sy = Math.hypot(r(0, 0), r(1, 0));
  if (sy > 1e-6) {
    return {
      roll: Math.atan2(r(1, 0), r(0, 0)),
      yaw: Math.atan2(-r(2, 0), sy),
      pitch: Math.atan2(r(2, 1), r(2, 2)),
    };
  }
  return { roll: 0, yaw: Math.atan2(-r(2, 0), sy), pitch: Math.atan2(-r(1, 2), r(1, 1)) };
}

export function signalsFromResult(result) {
  // Build a per-frame signal from a FaceLandmarker result, or null if no face.
  if (!result.faceLandmarks || result.faceLandmarks.length === 0) return null;
  const lm = result.faceLandmarks[0];
  const [rfx, rfy] = eyeFeature(lm, RIGHT_EYE);
  const [lfx, lfy] = eyeFeature(lm, LEFT_EYE);
  const openness = (eyeOpenness(lm, RIGHT_EYE) + eyeOpenness(lm, LEFT_EYE)) / 2;
  let yaw = 0, pitch = 0, roll = 0;
  const mats = result.facialTransformationMatrixes;
  if (mats && mats.length) ({ yaw, pitch, roll } = headPoseFromMatrix(mats[0].data));
  return { fx: (rfx + lfx) / 2, fy: (rfy + lfy) / 2, yaw, pitch, roll, openness };
}

// Twelve design terms: a second-order polynomial in the iris feature, linear head
// yaw and pitch, and the products of head pose with the iris feature. The pose-by-
// feature products let head pose change the feature's gain, not only add an offset.
// That is what a still-head calibration held at two poses needs in order to fit
// both poses instead of averaging them. Roll is handled by the roll-invariant feature.
function design(s) {
  const { fx, fy, yaw, pitch } = s;
  return [
    1, fx, fy, fx * fx, fy * fy, fx * fy, yaw, pitch,
    yaw * fx, yaw * fy, pitch * fx, pitch * fy,
  ];
}

function median(values) {
  const ordered = [...values].sort((a, b) => a - b);
  const n = ordered.length, mid = n >> 1;
  return n % 2 ? ordered[mid] : (ordered[mid - 1] + ordered[mid]) / 2;
}

export function medianSignal(samples) {
  const field = (k) => median(samples.map((s) => s[k]));
  return {
    fx: field("fx"), fy: field("fy"), yaw: field("yaw"),
    pitch: field("pitch"), roll: field("roll"), openness: field("openness"),
  };
}

export class HeadAwareModel {
  constructor(coeffsX, coeffsY, poseLo, poseHi) {
    this.coeffsX = coeffsX;
    this.coeffsY = coeffsY;
    this.poseLo = poseLo;
    this.poseHi = poseHi;
  }

  static fit(signals, targets, ridge = 1e-3) {
    if (signals.length !== targets.length) {
      throw new Error("signals and targets must have the same length");
    }
    const X = signals.map(design);
    const coeffsX = ridgeFit(X, targets.map((t) => t[0]), ridge);
    const coeffsY = ridgeFit(X, targets.map((t) => t[1]), ridge);
    const axis = (k) => signals.map((s) => s[k]);
    const poseLo = [Math.min(...axis("yaw")), Math.min(...axis("pitch")), Math.min(...axis("roll"))];
    const poseHi = [Math.max(...axis("yaw")), Math.max(...axis("pitch")), Math.max(...axis("roll"))];
    return new HeadAwareModel(coeffsX, coeffsY, poseLo, poseHi);
  }

  map(s) {
    const t = design(s);
    const nx = this.coeffsX.reduce((acc, c, i) => acc + c * t[i], 0);
    const ny = this.coeffsY.reduce((acc, c, i) => acc + c * t[i], 0);
    return [nx, ny];
  }

  poseDriftDegrees(s) {
    const outside = (v, lo, hi) => Math.max(0, lo - v, v - hi);
    const worst = Math.max(
      outside(s.yaw, this.poseLo[0], this.poseHi[0]),
      outside(s.pitch, this.poseLo[1], this.poseHi[1]),
      outside(s.roll, this.poseLo[2], this.poseHi[2]),
    );
    return (worst * 180) / Math.PI;
  }
}

export function crossValidatedError(centres, targets, ridge = 1e-3, groups = null) {
  // Leave-one-group-out mean error in normalised screen units. Grouping by screen
  // target holds out all of a target's poses together, so no pose leaks.
  const n = centres.length;
  const labels = groups || centres.map((_, i) => i);
  const folds = [...new Set(labels)];
  if (folds.length < 3) throw new Error("need at least three groups for cross-validation");
  let total = 0, count = 0;
  for (const held of folds) {
    const trainSignals = [], trainTargets = [];
    for (let j = 0; j < n; j++) {
      if (labels[j] !== held) { trainSignals.push(centres[j]); trainTargets.push(targets[j]); }
    }
    const model = HeadAwareModel.fit(trainSignals, trainTargets, Math.max(ridge, 1e-6));
    for (let j = 0; j < n; j++) {
      if (labels[j] === held) {
        const [px, py] = model.map(centres[j]);
        total += Math.hypot(px - targets[j][0], py - targets[j][1]);
        count += 1;
      }
    }
  }
  return total / count;
}

File: src/readsync/webcam/gaze.py
"""Head-pose-aware gaze model for the webcam strand.

The iris feature alone is confounded by head orientation, so even a good
polynomial calibration drifts when the head turns. This model adds the head yaw
and pitch as inputs, so the calibration can compensate for the head pose the
reader actually held, and it records how far the head has moved from the
calibration pose, so gaze can be flagged when the reader has shifted too far to
trust.

The fit is a regularised linear least squares over an explicit design that mixes a
second-order polynomial in the iris feature with head-pose terms and their products
with the feature, so head pose can rescale the feature, not only shift it.
The maths is pure Python and is tested; only reading the signals from a frame needs
a camera.
"""

from __future__ import annotations

import math
from collections.abc import Sequence
from dataclasses import dataclass

from ._linalg import ridge_fit
from .tracker import GazeSignals

__all__ = [
    "HeadAwareGazeModel",
    "aggregate_signals",
    "fit_head_aware",
    "cross_validated_error",
]

# Twelve design terms: a second-order polynomial in the iris feature, linear head
# yaw and pitch, and the products of head pose with the iris feature. The pose-by-
# feature products let head pose change the gain of the feature, not only add an
# offset, which is what a still-head calibration held at two poses needs in order to
# fit both poses instead of averaging them. Roll is handled by the roll-invariant
# iris feature itself.
_N_TERMS = 12


def _design(signals: GazeSignals) -> list[float]:
    fx, fy = signals.fx, signals.fy
    yaw, pitch = signals.yaw, signals.pitch
    return [
        1.0, fx, fy, fx * fx, fy * fy, fx * fy, yaw, pitch,
        yaw * fx, yaw * fy, pitch * fx, pitch * fy,
    ]


def _median(values: Sequence[float]) -> float:
    ordered = sorted(values)
    n = len(ordered)
    mid = n // 2
    return ordered[mid] if n % 2 else (ordered[mid - 1] + ordered[mid]) / 2


def aggregate_signals(samples: Sequence[GazeSignals]) -> GazeSignals:
    """Reduce a target's signal samples to their per-field median.

    The median ignores the occasional stray frame, for example a half-blink during
    a fixation, so it is a steadier centre than the mean.
    """
    if not samples:
        raise ValueError("need at least one signal sample to aggregate")
    return GazeSignals(
        _median([s.fx for s in samples]),
        _median([s.fy for s in samples]),
        _median([s.yaw for s in samples]),
        _median([s.pitch for s in samples]),
        _median([s.roll for s in samples]),
        _median([s.openness for s in samples]),
    )


@dataclass(frozen=True)
class HeadAwareGazeModel:
    """Maps gaze signals to a normalised screen position, compensating head pose.

    ``pose_lo`` and ``pose_hi`` are the per-axis range of head orientation seen
    during calibration. Calibrating at more than one head position widens that
    range and gives the yaw and pitch terms something to learn from, so the model
    compensates across the calibrated range, not only near a single pose.
    """

    coeffs_x: list[float]
    coeffs_y: list[float]
    pose_lo: tuple[float, float, float]
    pose_hi: tuple[float, float, float]

    @classmethod
    def fit(
        cls,
        signals: Sequence[GazeSignals],
        targets: Sequence[tuple[float, float]],
        *,
        ridge: float = 1e-3,
    ) -> HeadAwareGazeModel:
        """Fit the model from per-target signals to known screen targets."""
        if len(signals) != len(targets):
            raise ValueError("signals and targets must have the same length")
        if len(signals) < _N_TERMS and ridge <= 0:
            raise ValueError(f"need at least {_N_TERMS} calibration points, or set ridge > 0")
        design = [_design(s) for s in signals]
        coeffs_x = ridge_fit(design, [tx for tx, _ in targets], ridge)
        coeffs_y = ridge_fit(design, [ty for _, ty in targets], ridge)
        yaws = [s.yaw for s in signals]
        pitches = [s.pitch for s in signals]
        rolls = [s.roll for s in signals]
        pose_lo = (min(yaws), min(pitches), min(rolls))
        pose_hi = (max(yaws), max(pitches), max(rolls))
        return cls(coeffs_x, coeffs_y, pose_lo, pose_hi)

    def map(self, signals: GazeSignals) -> tuple[float, float]:
        """Predict the normalised screen position for one frame's signals."""
        terms = _design(signals)
        nx = sum(c * v for c, v in zip(self.coeffs_x, terms, strict=True))
        ny = sum(c * v for c, v in zip(self.coeffs_y, terms, strict=True))
        return nx, ny

    def pose_drift_degrees(self, signals: GazeSignals) -> float:
        """How far the head is outside the calibrated pose range, in degrees.

        Zero while the head stays within the range seen at calibration, growing
        once it moves beyond it, which is where the compensation stops being
        supported by data and the gaze should no longer be trusted.
        """

        def outside(value: float, low: float, high: float) -> float:
            return max(0.0, low - value, value - high)

        worst = max(
            outside(signals.yaw, self.pose_lo[0], self.pose_hi[0]),
            outside(signals.pitch, self.pose_lo[1], self.pose_hi[1]),
            outside(signals.roll, self.pose_lo[2], self.pose_hi[2]),
        )
        return math.degrees(worst)


def fit_head_aware(
    per_target_samples: Sequence[Sequence[GazeSignals]],
    targets: Sequence[tuple[float, float]],
    *,
    ridge: float = 1e-3,
) -> HeadAwareGazeModel:
    """Aggregate each target's samples, then fit a head-aware model."""
    if len(per_target_samples) != len(targets):
        raise ValueError("samples and targets must have the same length")
    centres = [aggregate_signals(samples) for samples in per_target_samples]
    return HeadAwareGazeModel.fit(centres, targets, ridge=ridge)


def cross_validated_error(
    centres: Sequence[GazeSignals],
    targets: Sequence[tuple[float, float]],
    *,
    ridge: float = 1e-3,
    groups: Sequence[int] | None = None,
) -> float:
    """Leave-one-group-out mean error in normalised screen units.

    For each group the model is fitted on the others and then used to predict the
    held-out group's points. The mean predicted-to-true distance is an unbiased
    accuracy estimate, because no point is scored by a model that saw it. With
    multi-pose calibration, ``groups`` labels each point with its screen target, so
    a target's samples at every head pose are held out together and a pose seen for
    that target elsewhere cannot leak in. Without ``groups`` each point is its own
    group, the plain leave-one-out case.
    """
    n = len(centres)
    if n != len(targets):
        raise ValueError("centres and targets must have the same length")
    labels = list(range(n)) if groups is None else list(groups)
    if len(labels) != n:
        raise ValueError("groups must have one label per point")
    folds = sorted(set(labels))
    if len(folds) < 3:
        raise ValueError("need at least three groups for cross-validation")
    total = 0.0
    count = 0
    for held in folds:
        train_signals = [centres[j] for j in range(n) if labels[j] != held]
        train_targets = [targets[j] for j in range(n) if labels[j] != held]
        model = HeadAwareGazeModel.fit(train_signals, train_targets, ridge=max(ridge, 1e-6))
        for j in range(n):
            if labels[j] == held:
                px, py = model.map(centres[j])
                tx, ty = targets[j]
                total += math.hypot(px - tx, py - ty)
                count += 1
    return total / count
